### PR TITLE
CDRIVER-3905 Change estimatedDocumentCount to use collStats on 4.9+

### PR DESCRIPTION
**Pull request by benjirewis**
[CDRIVER-3905](https://jira.mongodb.org/browse/CDRIVER-3905)

Test changes in first three commits, code changes in remaining three.

Updates `mongoc_collection_estimated_document_count` to use an aggregate with the $collStats pipeline stage when running against server version 4.9 or higher. Adds a new wire version macro for wire version 12.

Adds new spec tests for estimatedDocumentCount under `crud/unified` and moves old CRUD tests to `crud/legacy`. Updates unified test runner accordingly. Updates retryable reads and versioned API spec tests involving estimatedDocumentCount.

### CODE REVIEW

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [inline src/libmongoc/src/mongoc/mongoc-collection.c:918]: Although this swallows the error and returns 0, it does not touch the passed-in `reply` which still contains the NamespaceNotFound error. This is problematic, as users will not expect the reply to contain an error when the returned `error` is NULL. I could instead destroy `reply` and document that `reply` may be NULL when running against a non-existent collection. I could also mock a server response in `reply`, but mocking server behavior client-side sounds like trouble. 

@alcaeus I'm curious what would make more sense in PHP.

File: src/libmongoc/tests/unified/result.c
Comment (benjirewis) [inline src/libmongoc/tests/unified/result.c:356]: This was an existing bug in the unified test runner. We should not necessarily be expecting the result to be ok when `expectEvents` are specified according to the [spec](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectedevent). I'll handle this change through a separate PR and ticket and rebase this PR eventually.

File: src/libmongoc/tests/bsonutil/bson-parser.c
Comment (benjirewis) [inline src/libmongoc/tests/bsonutil/bson-parser.c:642]: This was also an existing bug in the unified test runner. `bson_parser_parse` is sometimes called when the passed in document `in` is NULL, and it will cause an error in `BSON_FOREACH` if unchecked. I'll handle this change through a separate PR and ticket and rebase this PR eventually.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-collection.c:918]: PHP does not use the CRUD API from libmongoc but implements its own in the PHP library. As such, we're not affected by changes to this method.

Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-collection.c:918]: Upon taking a closer look, we'll have to consider the server reply in general: even when the collection exists, the reply will be different from what a user might expect to be returned. I'm not sure how much of this is covered by a BC promise, but users won't see the usual `count` reply document when running on 4.9+.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:916]: I may be mistaken, but I believe that this will cause memory to leak as we're not resetting error information, but rather setting the pointer to `NULL`. This should either use `bson_set_error` or `memset` to make sure there's no error information.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:923]: If the command was successful but we don't find the data we're looking for, we're returning a count of `-1` without exposing any additional information. It may be sensible to set `error` to something that indicates that we didn't receive the information we were looking for.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:906]: I mentioned this in the other comment and want to leave the final decision to @kevinAlbs and @bazile-clyde, but we may want to consider crafting a `reply` that matches the `count` command reply document when using agg pipeline to determine document count. While technically the details of how we get to a document count are internal and shouldn't be relied on by users, the fact that the reply document changes significantly may present a problem for some users.

I don't think we need to do anything for the error case, but in case we're returning any count we may want to craft a document like `{ "n" : <count>, "ok" : 1 }` to return when using agg pipeline to retrieve document count.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:850]: Using the `BSON_APPEND_` macros would make this a little easier to read as it removes the need to pass string lengths for everything. Compare `bson_append_utf8` for example:
```
bson_append_utf8 (out, "aggregate", 9, coll->collection, coll->collectionlen);

BSON_APPEND_UTF8 (out, "aggregate", coll->collection);
```

Depending on whether clang-format allows you to do so,  leveraging indentation could make this a little more readable as well:
```
BSON_APPEND_UTF8 (out, "aggregate", coll->collection);

[...]

BSON_APPEND_DOCUMENT_BEGIN (&pipeline, keys[key++], &coll_stats_stage);
   BSON_APPEND_DOCUMENT_BEGIN (&coll_stats_stage, "$collStats", &coll_stats_stage_doc);
      BSON_APPEND_DOCUMENT_BEGIN (&coll_stats_stage_doc, "count", &empty);
      bson_append_document_end (&coll_stats_stage_doc, &empty);
   bson_append_document_end (&coll_stats_stage, &coll_stats_stage_doc);
bson_append_document_end (&pipeline, &coll_stats_stage);
```

This could help make sense of the nesting a little better, but as I said, not sure if clang-format will allow you to do this.

**Review by alcaeus — COMMENTED** [review]
Have a couple of notes, but the changes LGTM in general. I'd like to let @bazile-clyde or @kevinAlbs chime in on the reply and error pointer comments once they're available.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:850]: Awesome! I wasn't aware of these macros. I've added them in but unfortunately cannot do the indentation you described. As you predicted, clang-format will not allow indentation of regular statements like this.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (bazile-clyde) [reply src/libmongoc/src/mongoc/mongoc-collection.c:916]: I'm not sure if this a memory leak but I'm sure it's a no-op. Because C passes everything by value, this line changes a copy of the parameter's address and not the actual parameter itself or its contents. If this was passed by reference somehow then this would be a memory leak. Here's a code snippet of what's happening:
```
void faux_mod(int *ptr) {
	ptr = NULL;
	printf("faux_mod: address=%p\n", ptr ? ptr : 0);
}

void real_mod(int *ptr) {
	*ptr = 2;
	printf("real_mod: address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);
}

int main() {
   int* ptr;
   ptr = (int*) malloc(sizeof(int));
   *ptr = 1;

   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   faux_mod(ptr);
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   real_mod(ptr);
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   printf("set to NULL\n");
   ptr = NULL;
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   free(ptr);
}

...
# output
address=0x558bcab3e2a0,val=1
faux_mod: address=(nil)
address=0x558bcab3e2a0,val=1
real_mod: address=0x558bcab3e2a0,val=2
address=0x558bcab3e2a0,val=2
set to NULL
address=(nil),val=0

```

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:906]: It's a great point, and it seems like we should put some thought into the consistency of the returned `reply`. I've created [CDRIVER-3933](https://jira.mongodb.org/browse/CDRIVER-3933) to do this work separately from this PR. The spec only requires us to return an int64 `count` and makes no mention of returning the server response, so we may get rid of `reply` entirely.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:916]: Yep, you're totally right! Very helpful explanation, thank you @bazile-clyde . I only use `bson_set_error` now.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:918]: As mentioned in the other thread, [CDRIVER-3933](https://jira.mongodb.org/browse/CDRIVER-3933) will be used for considering the future of the estimated document count `reply`.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:923]: An interesting point! The estimated document count function did not return an error on a successful but malformed server response before this added functionality, so I'm not sure what a meaningful error would be. Any thoughts @bazile-clyde ?

File: src/libmongoc/tests/bsonutil/bson-parser.c
Comment (benjirewis) [reply src/libmongoc/tests/bsonutil/bson-parser.c:642]: Done.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:835]: Nit: I feel like the `keys` variable doesn't bring any additional value:
```suggestion
   BSON_APPEND_DOCUMENT_BEGIN (&pipeline, "0", &coll_stats_stage);
```

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:843]: Same nit as above:
```suggestion
   BSON_APPEND_DOCUMENT_BEGIN (&pipeline, "1", &group_stage);
```

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [inline src/libmongoc/src/mongoc/mongoc-collection.c:826]: Can be dropped if you decide to accept the suggestions below:
```suggestion
```

**Review by alcaeus — APPROVED** [review]
LGTM besides nits I couldn't help myself with.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-collection.c:907]: Sounds good to me, thanks for creating the follow-up ticket!

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-collection.c]: Since this is an extremely narrow use-case (only thing I can imagine is the server not returning an initial batch in the cursor), I'm fine leaving this as is and revisiting as part of CDRIVER-3933 or another ticket.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (alcaeus) [reply src/libmongoc/src/mongoc/mongoc-collection.c]: Thanks for that explanation @bazile-clyde, I completely forgot about pass-by-value 🤦‍♂️

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:826]: Sounds good! Done.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:835]: Done.

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:843]: Done.

**Comment by alcaeus** [comment]
You're a victim of your own success - want to include changes from https://github.com/mongodb/specifications/pull/933 in this PR?

**Review by bazile-clyde — APPROVED** [review]
LGTM!

File: src/libmongoc/src/mongoc/mongoc-collection.c
Comment (benjirewis) [reply src/libmongoc/src/mongoc/mongoc-collection.c:914]: Ah that seems better! Thanks.

**Comment by benjirewis** [comment]
I can't remove the skipReason just yet, as the "latest" server version hasn't updated on Evergreen to include [this commit](https://evergreen.mongodb.com/version/mongodb_mongo_master_075e51fefecc54e219c56e75c3ba08c993b5c187). So, I'll do CDRIVER-3932 in a separate PR.

**Pull request merged**
